File: ta4j-core/src/main/java/org/ta4j/core/backtest/BacktestExecutorBuilder.java
package org.ta4j.core.backtest;

import org.ta4j.core.backtest.analysis.cost.CostModel;
import org.ta4j.core.backtest.analysis.cost.ZeroCostModel;
import org.ta4j.core.num.DoubleNumFactory;
import org.ta4j.core.num.NumFactory;

/**
 * @author Lukáš Kvídera
 */
public class BacktestExecutorBuilder {

  private NumFactory numFactory = DoubleNumFactory.getInstance();
  /** The trading cost models */
  private CostModel transactionCostModel = new ZeroCostModel();
  private CostModel holdingCostModel = new ZeroCostModel();

  /** The trade execution model to use */
  private TradeExecutionModel tradeExecutionModel = new TradeOnCurrentCloseModel();


  public BacktestExecutorBuilder numFactory(final NumFactory numFactory) {
    this.numFactory = numFactory;
    return this;
  }


  public BacktestExecutorBuilder transactionCostModel(final CostModel transactionCostModel) {
    this.transactionCostModel = transactionCostModel;
    return this;
  }


  public BacktestExecutorBuilder holdingCostModel(final CostModel holdingCostModel) {
    this.holdingCostModel = holdingCostModel;
    return this;
  }


  public BacktestExecutorBuilder tradeExecutionModel(final TradeExecutionModel tradeExecutionModel) {
    this.tradeExecutionModel = tradeExecutionModel;
    return this;
  }


  public BacktestExecutor build() {
    return new BacktestExecutor(
        new BacktestConfiguration(
            this.numFactory,
            this.transactionCostModel,
            this.holdingCostModel,
            this.tradeExecutionModel
        )
    );
  }
}

File: ta4j-core/src/main/java/org/ta4j/core/backtest/BacktestExecutorBuilder.kt
package org.ta4j.core.backtest

import org.ta4j.core.backtest.analysis.cost.CostModel
import org.ta4j.core.backtest.analysis.cost.ZeroCostModel
import org.ta4j.core.num.DoubleNumFactory
import org.ta4j.core.num.NumFactory
import org.ta4j.core.num.NumFactoryProvider.defaultNumFactory

/**
 * @author Lukáš Kvídera
 */
class BacktestExecutorBuilder {
    private var numFactory: NumFactory = DoubleNumFactory

    /** The trading cost models  */
    private var transactionCostModel: CostModel = ZeroCostModel
    private var holdingCostModel: CostModel = ZeroCostModel

    /** The execution mode to use  */
    private var executionMode: ExecutionMode = ExecutionMode.CURRENT_CLOSE


    fun numFactory(numFactory: NumFactory): BacktestExecutorBuilder = apply {
        this.numFactory = numFactory
        defaultNumFactory = numFactory
    }


    fun transactionCostModel(transactionCostModel: CostModel): BacktestExecutorBuilder = apply {
        this.transactionCostModel = transactionCostModel
    }


    fun holdingCostModel(holdingCostModel: CostModel): BacktestExecutorBuilder = apply {
        this.holdingCostModel = holdingCostModel
    }


    fun executionMode(executionMode: ExecutionMode): BacktestExecutorBuilder = apply {
        this.executionMode = executionMode
    }


    fun build(): BacktestExecutor = BacktestExecutor(
        BacktestConfiguration(
            this.numFactory,
            this.transactionCostModel,
            this.holdingCostModel,
            this.executionMode
        )
    )
}
